aten/src/ATen/core/dispatch/Dispatcher.cpp

```diff
@@ -254,6 +255,66 @@ void Dispatcher::deregisterDef_(
   cleanup(op, op_name);
 }
 
+namespace {
+
+using AbstractImplPyStubsType = std::unordered_map<at::OperatorName, const char*>;
+AbstractImplPyStubsType& kAbstractImplPyStubs() {
+  static AbstractImplPyStubsType _data;
+  return _data;
+}
+
+}
+
+RegistrationHandleRAII Dispatcher::registerAbstractImplPyStub(
+  OperatorName op_name,
+  const char* pymodule
+) {
+  std::lock_guard<std::mutex> lock(guard_->mutex);
+  // If there are duplicates, we just let it through and warn about it.
+  // Throwing an error during static initialization causes a crash that
+  // doesn't give any sign of what happened.
+  if (kAbstractImplPyStubs().find(op_name) != kAbstractImplPyStubs().end()) {
+    TORCH_WARN(
+        "Tried to register an abstract impl pystub for ", op_name, " ",
+        "but there already was one. We will override the existing pystub.");
+  }
+  kAbstractImplPyStubs()[op_name] = pymodule;
+  return RegistrationHandleRAII([guard = this->guard_, op_name] {
+    std::lock_guard<std::mutex> lock(guard->mutex);
+    if (!guard->alive.load()) {
+      return;
+    }
+    kAbstractImplPyStubs().erase(op_name);
+  });
+}
+
+bool Dispatcher::maybeImportAbstractImplPyStub(OperatorName op_name) {
+  const char* pymodule = nullptr;
+  {
+    std::lock_guard<std::mutex> lock(guard_->mutex);
+    auto elt = kAbstractImplPyStubs().find(op_name);
+    if (elt == kAbstractImplPyStubs().end()) {
+      return false;
+    }
+    pymodule = elt->second;
+    // We need to release the lock here because the pyimport will register things to
+    // the dispatcher and that needs to grab the lock. It's possible that someone
+    // will unload the shared library that owns the existing impl_abstract_pystub,
+    // or load a library with a different impl_abstract_pystub for the same operator,
+    // but those will cause other problems so we don't worry about it here.
+  }
+  auto* interpreter = at::impl::PythonOpRegistrationTrampoline::getInterpreter();
+  TORCH_CHECK(
+      interpreter != nullptr,
+      op_name,
+      ": while attempting to run this operator with Meta Tensors: "
+      "the abstract impl for this operator (necessary for Meta Tensors) "
+      "was declared to exist in the Python module ", pymodule,
+      " but Python is not available.");
+  (*interpreter)->op_registration_pyimport(pymodule, toString(op_name));
```

Comment:
So in the event of a race, this can trigger multiple times right? (Which should be OK because module imports should be idempotent)

Reply:
I had to check, but it looks like the dispatcher is read-write thread safe, which is *phew*

Reply:
Actually, no, this is kind of sus. Only *part* of the dispatcher is read-write threadsafe, and not the important part.

The critical data structure to protect is

```
std::array<KernelFunction, c10::num_runtime_entries> dispatchTable_;
```

which is what actually gets poked when we actually attempt to do an operator call. The race is when you have two threads simultaneously trying to trigger a meta call: one thread will trigger the pyimport load while the other thread is attempting to do lookup.

KernelFunction is a chonky boy

```
  BoxedKernel boxed_kernel_func_;
  void* unboxed_kernel_func_;
  void* sym_unboxed_kernel_func_;
```

which means that even in x86 TSO semantics it cannot be atomically updated. So you could potentially see a partially invalid KernelFunction in concurrent access.

This problem occurs solely because we are relying on the imported module to directly overwrite the dispatcher entry for Meta upon import, which is the logical and BC way to do things. Can we easily make this thread safe? In general we cannot take out a lock on reads for generic kernels.

One potential fix is to prevent `impl_abstract` from actually registering into the dispatch table directly, but instead some side table that is read-write race safe. The meta fallback redispatches into that table when the entry exists.

Or lmk if you want to YOLO, since in practice you're likely to be guarded by a higher level lock already, the torch.compile lock.

Reply:
A race could happen if multiple threads invoke DispatchKey::Meta's fallback (so `op_registration_pyimport` could get run by multiple threads). op_registration_pyimport needs to acquire the GIL, so the GIL would make it so that only one thread modifies the Dispatcher state at a time. (also yes, imports are idempotent, so the second time an import happens it no-ops).

Reply:
Oh I see, you're right. Let me think about this for a bit

Reply:
Oh, an even simpler way to do this is to just forbid Meta registrations (or any other registrations) here (only go into the fake tensor side table). Because I don't think we need the *meta* for anything.